src/Native/Unix/System.Security.Cryptography.Native.Apple/pal_rsa.cpp

```diff
@@ -0,0 +1,258 @@
+{
+    if (pPublicKey == nullptr || pPrivateKey == nullptr || pOSStatus == nullptr)
+        return kErrorBadInput;
+    if (keySizeBits < 384 || keySizeBits > 16384)
+        return -2;
+
+    *pPublicKey = nullptr;
+    *pPrivateKey = nullptr;
+
+    CFMutableDictionaryRef attributes = CFDictionaryCreateMutable(nullptr, 2, &kCFTypeDictionaryKeyCallBacks, nullptr);
+
+    CFNumberRef cfKeySizeValue = CFNumberCreate(nullptr, kCFNumberIntType, &keySizeBits);
+
+    CFDictionaryAddValue(attributes, kSecAttrKeyType, kSecAttrKeyTypeRSA);
+    CFDictionaryAddValue(attributes, kSecAttrKeySizeInBits, cfKeySizeValue);
+
+    OSStatus status = SecKeyGeneratePair(attributes, pPublicKey, pPrivateKey);
+
+    CFRelease(attributes);
+    CFRelease(cfKeySizeValue);
+
+    *pOSStatus = status;
+    return status == noErr;
+}
+
+static int32_t ExecuteOaepTransform(SecTransformRef xform,
+                                    uint8_t* pbData,
+                                    int32_t cbData,
+                                    PAL_HashAlgorithm algorithm,
+                                    CFDataRef* pDataOut,
+                                    CFErrorRef* pErrorOut)
+{
+    int ret = INT_MIN;
+
+    if (!SecTransformSetAttribute(xform, kSecPaddingKey, kSecPaddingOAEPKey, pErrorOut))
+    {
+        ret = kErrorSeeError;
+        goto cleanup;
+    }
+
+    // Documentation mentions kSecOAEPMGF1DigestAlgorithmAttributeName, but on the Apple platform
+    // "SHA2" is an algorithm and the size is encoded separately. Since there doesn't seem to be
+    // a second attribute to encode SHA2-256 vs SHA2-384, be limited to SHA-1.
+    if (algorithm != PAL_SHA1)
+    {
+        ret = kErrorUnknownAlgorithm;
+        goto cleanup;
+    }
+
+    ret = ExecuteCFDataTransform(xform, pbData, cbData, pDataOut, pErrorOut);
+
+cleanup:
+    return ret;
+}
+
+extern "C" int32_t AppleCryptoNative_RsaDecryptOaep(SecKeyRef privateKey,
+                                                    uint8_t* pbData,
+                                                    int32_t cbData,
+                                                    PAL_HashAlgorithm mfgAlgorithm,
+                                                    CFDataRef* pDecryptedOut,
+                                                    CFErrorRef* pErrorOut)
+{
+    if (privateKey == nullptr || pbData == nullptr || cbData < 0 || pDecryptedOut == nullptr || pErrorOut == nullptr)
+    {
+        return kErrorBadInput;
+    }
+
+    *pDecryptedOut = nullptr;
+    *pErrorOut = nullptr;
+
+    int ret = INT_MIN;
+    SecTransformRef decryptor = SecDecryptTransformCreate(privateKey, pErrorOut);
+
+    if (decryptor == nullptr || *pErrorOut != nullptr)
+    {
+        ret = kErrorSeeError;
+        goto cleanup;
+    }
+
+    ret = ExecuteOaepTransform(decryptor, pbData, cbData, mfgAlgorithm, pDecryptedOut, pErrorOut);
+
+cleanup:
+
+    if (decryptor != nullptr)
+    {
+        CFRelease(decryptor);
+    }
+
+    return ret;
```

Comment:
The nesting might get pretty deep, but I'll look into it.  It was definitely a bit of "I'm writing native code, so that means there's a cleanup label and a bunch of gotos".